aggregator/src/aggregator/report_writer.rs

```diff
@@ -298,18 +318,76 @@ where
                 }
             }
             Err(error) => {
-                if let Error::ReportRejected(_rejection) = error {
-                    // Incrementing task counters disabled for now
-                    // https://github.com/divviup/janus/issues/2654
-                    // tx.increment_task_upload_counter(
-                    //     rejection.task_id(),
-                    //     ord,
-                    //     &rejection.reason().into(),
-                    // )
-                    // .await?;
+                if let Error::ReportRejected(rejection) = error {
+                    task_upload_counter.increment_report_rejection(&rejection);
                 }
                 Err(error)
             }
         }
     }
 }
+
+/// A collection of [`TaskUploadCounter`]s.
+#[derive(Debug)]
+pub struct TaskUploadCounters(StdMutex<BTreeMap<TaskId, TaskUploadCounter>>);
```

Comment:
Hmm, [usage of this struct would be having it wrapped the struct itself wrapped in an Arc](https://github.com/divviup/janus/pull/2684/files#diff-cb4c437780754977b18c6df75c86687bd6fd7bda43e46a76946d4d145ea010a1R182), to share it across futures. But it might be less surprising and more idiomatic to wrap the Mutex in an Arc in the struct definition and have the struct derive `Clone`. Then usage can just clone the struct to share it across futures.

Reply:
Oh, I see, that makes sense. It's kind of six of one, half dozen of the other, but I feel like putting the `Arc` inside of `TaskUploadCounters` is a little more resistant to misuse. Really, I think that we should either have `TaskUploadCounters(Arc<StdMutex<BTree<etc.>>>);` and document that callers can `clone` it across threads, or have `TaskUploadCounters(BTree<etc.>)` and document that callers should wrap it in `Arc<Mutex<>>` to make it threadsafe. It's awkward having the struct implement half a synchronization story.

And the locking logic is best hidden behind `TaskUploadCounters`' methods, so I think moving the `Arc` inside the struct is the right move.

Reply:
One way to avoid the `into_inner` which must succeed during `write` would be to:
1. Take the lock.
2. `mem::swap` an empty `BTreeMap` into the mutex's storage, taking ownership of the filled `BTreeMap`. (creating a new, empty `BTreeMap` doesn't allocate)
3. Drop the lock.
4. Continue the logic as previously, using the `BTreeMap` swapped out of the mutex to provide values to write back to the datastore.